In [4]:
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

import pandas as pd
import seaborn as sns

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error , r2_score
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LinearRegression
from sklearn.pipeline import Pipeline # 여러작업을 결합해 하나의 작업으로 생성
import numpy as np

In [5]:
df = pd.read_csv('./data/final3.csv')
df

rent_cnt      dust      temp  road_cnt  road_len      rain  강남구  강동구  \
0      -0.671517 -0.220561 -1.792863  0.088742  0.723189 -0.185544    0    0   
1      -0.502717 -0.062739 -1.869878  0.088742  0.723189 -0.185544    0    0   
2      -0.615250  0.068780 -1.927639  0.088742  0.723189 -0.185544    0    0   
3      -0.671517  0.147691 -1.725476  0.088742  0.723189 -0.185544    0    0   
4      -0.446450 -0.273169 -1.918012  0.088742  0.723189 -0.185544    0    0   
...          ...       ...       ...       ...       ...       ...  ...  ...   
794401 -0.840317 -0.167954 -2.120175  3.058366  1.869792 -0.185544    0    0   
794402 -0.755917  0.331817 -1.783237  3.058366  1.869792 -0.185544    0    0   
794403 -0.699651  0.594854 -1.350030  3.058366  1.869792 -0.185544    0    0   
794404 -0.671517 -0.273169 -1.763983  3.058366  1.869792 -0.185544    0    0   
794405 -0.784051 -0.430991 -2.033534  3.058366  1.869792 -0.185544    0    0   

        강북구  강서구  ...  맑음  비  폭우  평일  주말  공휴일  봄  여름  가을  겨울  
0         0    0  ...   1  0   0   0   0    1  0   0   0   1  
1         0    0  ...   1  0   0   0   1    0  0   0   0   1  
2         0    0  ...   1  0   0   0   1    0  0   0   0   1  
3         0    0  ...   1  0   0   1   0    0  0   0   0   1  
4         0    0  ...   1  0   0   1   0    0  0   0   0   1  
...     ...  ...  ...  .. ..  ..  ..  ..  ... ..  ..  ..  ..  
794401    0    0  ...   1  0   0   1   0    0  0   0   0   1  
794402    0    0  ...   1  0   0   1   0    0  0   0   0   1  
794403    0    0  ...   0  1   0   1   0    0  0   0   0   1  
794404    0    0  ...   1  0   0   1   0    0  0   0   0   1  
794405    0    0  ...   1  0   0   1   0    0  0   0   0   1  

[794406 rows x 44 columns]

In [3]:
# 상태 열 제거
df.drop(['상태'], axis=1, inplace=True)
df

대여소번호          날짜  대여건수.x  자치구  미세먼지  강수량.x  날씨  요일_그룹화  평균기온   구간합계  \
1         101  2021-01-01       7  마포구    30    0.0  맑음       3  -4.2   57.0   
2         101  2021-01-02      13  마포구    36    0.0  맑음       2  -5.0   57.0   
3         101  2021-01-03       9  마포구    41    0.0  맑음       2  -5.6   57.0   
4         101  2021-01-04       7  마포구    44    0.0  맑음       1  -3.5   57.0   
5         101  2021-01-05      15  마포구    28    0.0  맑음       1  -5.5   57.0   
...       ...         ...     ...  ...   ...    ...  ..     ...   ...    ...   
794402   4869  2021-12-27       1  송파구    32    0.0  맑음       1  -7.6  186.0   
794403   4869  2021-12-28       4  송파구    51    0.0  맑음       1  -4.1  186.0   
794404   4869  2021-12-29       6  송파구    61    0.2   비       1   0.4  186.0   
794405   4869  2021-12-30       7  송파구    28    0.0  맑음       1  -3.9  186.0   
794406   4869  2021-12-31       3  송파구    22    0.0  맑음       1  -6.7  186.0   

        길이합계  강수량.y  
1       65.7    0.0  
2       65.7    NaN  
3       65.7    NaN  
4       65.7    0.0  
5       65.7    NaN  
...      ...    ...  
794402  97.1    0.0  
794403  97.1    0.0  
794404  97.1    0.0  
794405  97.1    0.0  
794406  97.1    0.0  

[794406 rows x 12 columns]

In [4]:
# 컬럼명 변경
df.rename(columns={'대여건수.x':'대여건수'}, inplace=True)
df.rename(columns={'강수량.y':'강수량'}, inplace=True)
df

대여소번호          날짜  대여건수  자치구  미세먼지  강수량.x  날씨  요일_그룹화  평균기온   구간합계  \
1         101  2021-01-01     7  마포구    30    0.0  맑음       3  -4.2   57.0   
2         101  2021-01-02    13  마포구    36    0.0  맑음       2  -5.0   57.0   
3         101  2021-01-03     9  마포구    41    0.0  맑음       2  -5.6   57.0   
4         101  2021-01-04     7  마포구    44    0.0  맑음       1  -3.5   57.0   
5         101  2021-01-05    15  마포구    28    0.0  맑음       1  -5.5   57.0   
...       ...         ...   ...  ...   ...    ...  ..     ...   ...    ...   
794402   4869  2021-12-27     1  송파구    32    0.0  맑음       1  -7.6  186.0   
794403   4869  2021-12-28     4  송파구    51    0.0  맑음       1  -4.1  186.0   
794404   4869  2021-12-29     6  송파구    61    0.2   비       1   0.4  186.0   
794405   4869  2021-12-30     7  송파구    28    0.0  맑음       1  -3.9  186.0   
794406   4869  2021-12-31     3  송파구    22    0.0  맑음       1  -6.7  186.0   

        길이합계  강수량  
1       65.7  0.0  
2       65.7  NaN  
3       65.7  NaN  
4       65.7  0.0  
5       65.7  NaN  
...      ...  ...  
794402  97.1  0.0  
794403  97.1  0.0  
794404  97.1  0.0  
794405  97.1  0.0  
794406  97.1  0.0  

[794406 rows x 12 columns]

In [5]:
m_df = df

In [6]:
# 데이터 열 이름 수정
m_df.columns = ['station_num','date', 'rent_cnt','gu','dust','rain_seoul', 'weather','day','temp','road_cnt','road_len','rain_gu']
m_df.head()

station_num        date  rent_cnt   gu  dust  rain_seoul weather  day  \
1          101  2021-01-01         7  마포구    30         0.0      맑음    3   
2          101  2021-01-02        13  마포구    36         0.0      맑음    2   
3          101  2021-01-03         9  마포구    41         0.0      맑음    2   
4          101  2021-01-04         7  마포구    44         0.0      맑음    1   
5          101  2021-01-05        15  마포구    28         0.0      맑음    1   

   temp  road_cnt  road_len  rain_gu  
1  -4.2      57.0      65.7      0.0  
2  -5.0      57.0      65.7      NaN  
3  -5.6      57.0      65.7      NaN  
4  -3.5      57.0      65.7      0.0  
5  -5.5      57.0      65.7      NaN

In [7]:
# 결측치 채우기
m_df['rain_gu'].fillna(0, inplace=True)

In [8]:
print('데이터세트 Null 값 갯수 ', m_df.isnull().sum().sum())

데이터세트 Null 값 갯수  2145


In [9]:
m_df.isnull().sum()

station_num      0
date             0
rent_cnt         0
gu             715
dust             0
rain_seoul       0
weather          0
day              0
temp             0
road_cnt       715
road_len       715
rain_gu          0
dtype: int64

In [10]:
# 결측치가 있는 행은 삭제
m_df = m_df.dropna(how='any')

In [11]:
m_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 793691 entries, 1 to 794406
Data columns (total 12 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   station_num  793691 non-null  int64  
 1   date         793691 non-null  object 
 2   rent_cnt     793691 non-null  int64  
 3   gu           793691 non-null  object 
 4   dust         793691 non-null  int64  
 5   rain_seoul   793691 non-null  float64
 6   weather      793691 non-null  object 
 7   day          793691 non-null  int64  
 8   temp         793691 non-null  float64
 9   road_cnt     793691 non-null  float64
 10  road_len     793691 non-null  float64
 11  rain_gu      793691 non-null  float64
dtypes: float64(5), int64(4), object(3)
memory usage: 78.7+ MB


In [12]:
m_df.isnull().sum()

station_num    0
date           0
rent_cnt       0
gu             0
dust           0
rain_seoul     0
weather        0
day            0
temp           0
road_cnt       0
road_len       0
rain_gu        0
dtype: int64

In [6]:
data = df

In [14]:
from sklearn import preprocessing

# 레이블 인코딩 함수
def encode_features(dataDF):
    # 문자열 카테고리 피처
    features = ['date', 'weather', 'gu']
    for feature in features:
        le = preprocessing.LabelEncoder()
        le = le.fit(dataDF[feature])
        dataDF[feature] = le.transform(dataDF[feature])
        
    return dataDF

data = encode_features(data)
data.head()

C:\Users\mib91\AppData\Local\Temp/ipykernel_3244/1368731158.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataDF[feature] = le.transform(dataDF[feature])


station_num  date  rent_cnt  gu  dust  rain_seoul  weather  day  temp  \
1          101     0         7  12    30         0.0        0    3  -4.2   
2          101     1        13  12    36         0.0        0    2  -5.0   
3          101     2         9  12    41         0.0        0    2  -5.6   
4          101     3         7  12    44         0.0        0    1  -3.5   
5          101     4        15  12    28         0.0        0    1  -5.5   

   road_cnt  road_len  rain_gu  
1      57.0      65.7      0.0  
2      57.0      65.7      0.0  
3      57.0      65.7      0.0  
4      57.0      65.7      0.0  
5      57.0      65.7      0.0

In [15]:
# 요일과 관련된 df (data1)
data1 = data
data1

station_num  date  rent_cnt  gu  dust  rain_seoul  weather  day  temp  \
1               101     0         7  12    30         0.0        0    3  -4.2   
2               101     1        13  12    36         0.0        0    2  -5.0   
3               101     2         9  12    41         0.0        0    2  -5.6   
4               101     3         7  12    44         0.0        0    1  -3.5   
5               101     4        15  12    28         0.0        0    1  -5.5   
...             ...   ...       ...  ..   ...         ...      ...  ...   ...   
794402         4869   344         1  17    32         0.0        0    1  -7.6   
794403         4869   345         4  17    51         0.0        0    1  -4.1   
794404         4869   346         6  17    61         0.2        1    1   0.4   
794405         4869   347         7  17    28         0.0        0    1  -3.9   
794406         4869   348         3  17    22         0.0        0    1  -6.7   

        road_cnt  road_len  rain_gu  
1           57.0      65.7      0.0  
2           57.0      65.7      0.0  
3           57.0      65.7      0.0  
4           57.0      65.7      0.0  
5           57.0      65.7      0.0  
...          ...       ...      ...  
794402     186.0      97.1      0.0  
794403     186.0      97.1      0.0  
794404     186.0      97.1      0.0  
794405     186.0      97.1      0.0  
794406     186.0      97.1      0.0  

[793691 rows x 12 columns]

In [16]:
# 요일과 관련되지 않은 열 제거
# data1은 자치구를 고려 x
data1 = data1.drop(['date','station_num','rain_seoul','road_cnt', 'road_len', 'rain_gu', 'gu', 'weather', 'temp', 'dust'], axis=1)
data1

rent_cnt  day
1              7    3
2             13    2
3              9    2
4              7    1
5             15    1
...          ...  ...
794402         1    1
794403         4    1
794404         6    1
794405         7    1
794406         3    1

[793691 rows x 2 columns]

### 단순 선형회귀

In [7]:
# 특징과 타겟으로 나누기

x = data.drop(['rent_cnt'],axis=1)
y = data[['rent_cnt']]

In [8]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.model_selection import KFold

In [9]:
# 학습과 테스트 분리
X_train, X_test, y_train, y_test = train_test_split(x, y,
                                                   test_size=0.3, random_state=12)

In [10]:
from sklearn.linear_model import LinearRegression

# 빈 선형회귀 모델 생성
line_fitter = LinearRegression() 

# 학습 수행
line_fitter.fit(X_train, y_train)

LinearRegression()

In [11]:
# 학습이 완료된 모델 객체를 활용하여 예측 수행 
pred = line_fitter.predict(X_test)

In [12]:
from sklearn.metrics import mean_squared_error , r2_score

mse = mean_squared_error(y_test, pred)
rmse = np.sqrt(mse) # 사이킷런이 지원x

print('MSE : {0:.3f} , RMSE : {1:.3F}'.format(mse , rmse))
print('Variance score : {0:.3f}'.format(r2_score(y_test, pred)))

MSE : 0.812 , RMSE : 0.901
Variance score : 0.195


#### 회귀계수와 절편값 확인

In [10]:
print('절편값: ', line_fitter.intercept_)
print('회귀계수값', np.round(line_fitter.coef_, 1))

절편값:  [-1.8084341e+10]
회귀계수값 [[-0.00000000e+00  3.00000000e-01  4.37601594e+11 -3.32001512e+11
  -0.00000000e+00  4.36310958e+11  2.08016452e+11  2.48834778e+10
   1.20326991e+11  1.66238596e+11  6.36871622e+10 -6.15325314e+10
  -8.35029518e+10  1.17143363e+11  7.58016575e+10  1.56051513e+11
  -2.73859760e+09  2.22302813e+11 -9.64286324e+10  1.01698475e+10
   1.74699895e+10 -1.21494925e+10 -6.96535224e+11  7.84969032e+10
  -5.25750215e+10  1.38154255e+11  1.80298265e+11 -6.15325314e+10
   4.16039216e+09 -3.99460277e+10 -5.99642253e+10 -5.99642253e+10
  -5.99642253e+10  5.32125698e+10  5.32125698e+10  5.32125698e+10
   4.68580973e+09  4.68580973e+09  4.68580973e+09 -8.86534284e+08
  -8.86534284e+08 -8.86534284e+08 -8.86534284e+08]]


In [12]:
# 5개의 폴드 세트에 교차 검증 수행
# MSE와 RMSE 측정

from sklearn.model_selection import cross_val_score

y_target = data['rent_cnt']
X_data = data.drop(['rent_cnt'],axis=1,inplace=False)

lr = LinearRegression()

# 5 fold 셋으로 neg_MSE를 구함
neg_mse_scores = cross_val_score(lr, X_data, y_target, scoring='neg_mean_squared_error', cv=5)

In [13]:
neg_mse_scores

array([-0.32826718, -0.7144777 , -1.16254388, -1.48675965, -0.51163904])

In [14]:
# RMSE : neg-MSE에 -1  곱하고  sqrt()

rmse_scores = np.sqrt(-1*neg_mse_scores)
avg_rmse = np.mean(rmse_scores)

print(' 5 folds 의 개별 Negative MSE scores: ', np.round(neg_mse_scores, 2))
print(' 5 folds 의 개별 RMSE scores : ', np.round(rmse_scores, 2))
print(' 5 folds 의 평균 RMSE : {0:.3f} '.format(avg_rmse))

 5 folds 의 개별 Negative MSE scores:  [-0.33 -0.71 -1.16 -1.49 -0.51]
 5 folds 의 개별 RMSE scores :  [0.57 0.85 1.08 1.22 0.72]
 5 folds 의 평균 RMSE : 0.886 


### 다항 회귀

In [15]:
y_target = data['rent_cnt']
X_data = data.drop(['rent_cnt'],axis=1,inplace=False)

X_train , X_test , y_train , y_test = train_test_split(X_data , y_target ,test_size=0.3, random_state=156)

In [16]:
#(1) degree=2
# - 파이프라인 작업 구성
# Pipeline([('작업명',작업1), ('작업명',작업2),....('작업명',작업n)])
p_model = Pipeline([
    ('poly', PolynomialFeatures(degree=2)),
    ('linear', LinearRegression())
])

# pipeline에 data를 전달 해서 학습
p_model.fit(X_train, y_train)

Pipeline(steps=[('poly', PolynomialFeatures()), ('linear', LinearRegression())])

In [17]:
# linear스텝의 속성 coef_를 반환
print('Polynomial 회귀 계수\n', np.round(p_model.named_steps['linear'].coef_, 2))

Polynomial 회귀 계수
 [-4.86573560e+05 -2.03545967e+09 -7.67719275e+09  1.00808644e+09
  2.84323552e+09  4.11574860e+09 -1.15826118e+09 -3.49648864e+08
  1.64822188e+08 -4.27901391e+08  3.26394552e+07  2.19516349e+08
  2.26662382e+08  6.30030554e+07 -1.68176279e+08  1.64287681e+08
  1.25779010e+08 -9.56450471e+07 -1.52132230e+08 -4.02968600e+07
 -8.21824475e+07 -2.57063843e+07 -1.93033567e+07  6.82789050e+07
 -9.61660172e+07 -7.81663752e+07 -1.33525621e+07 -1.68692426e+07
  6.47668536e+07  7.66521540e+07  2.67059286e+07  4.72647056e+07
  2.01185004e+07  7.32968117e+07 -1.24463941e+08 -7.76043507e+06
 -6.79386617e+07 -1.95655918e+08  5.16384265e+06  1.62745121e+07
  1.20766633e+07  2.71444940e+05  2.59494896e+07 -6.39493345e+07
  0.00000000e+00  4.00000000e-02 -5.34184979e+07  1.80933665e+07
  0.00000000e+00  2.04684665e+07  3.35907398e+07  4.08962386e+06
  4.78996359e+07  3.87393933e+06  3.05649354e+07  2.92205591e+07
  1.74843152e+07  2.08445380e+07  1.31469285e+06  1.19169780e+07
  1.172

In [18]:
### 학습된 회귀식을 이용해 예측/평가
y_preds = p_model.predict(X_test)

# 평가
mse = mean_squared_error(y_test, y_preds)
rmse = np.sqrt(mse)

print('MSE : {0:.3f} , RMSE : {1:.3F}'.format(mse , rmse))
print('Variance score : {0:.3f}'.format(r2_score(y_test, y_preds)))

MSE : 0.789 , RMSE : 0.889
Variance score : 0.223


**다항회귀 degree=2 결과**
- MSE와 RMSE, variance score가 단항(선형)회귀보다 높게 나옴

In [19]:
#(2) degree=2, include_bias 튜닝
# - 파이프라인 작업 구성
# Pipeline([('작업명',작업1), ('작업명',작업2),....('작업명',작업n)])
p_model = Pipeline([
    ('poly', PolynomialFeatures(degree=2, include_bias=False)),
    ('linear', LinearRegression())
])

# pipeline에 data를 전달 해서 학습
p_model.fit(X_train, y_train)

### 학습된 회귀식을 이용해 예측/평가
y_preds = p_model.predict(X_test)

# 평가
mse = mean_squared_error(y_test, y_preds)
rmse = np.sqrt(mse)


print('MSE : {0:.3f} , RMSE : {1:.3F}'.format(mse , rmse))
print('Variance score : {0:.3f}'.format(r2_score(y_test, y_preds)))

MSE : 0.789 , RMSE : 0.889
Variance score : 0.223


In [20]:
## 3차 다항식으로 변경
# degree=3, include_bias=False
p_model = Pipeline([
    ('poly', PolynomialFeatures(degree=3, include_bias=False)),
    ('linear', LinearRegression())
])

# pipeline에 data를 전달 해서 학습
p_model.fit(X_train, y_train)

### 학습된 회귀식을 이용해 예측/평가
y_preds = p_model.predict(X_test)

# 평가
mse = mean_squared_error(y_test, y_preds)
rmse = np.sqrt(mse)


print('MSE : {0:.3f} , RMSE : {1:.3F}'.format(mse , rmse))
print('Variance score : {0:.3f}'.format(r2_score(y_test, y_preds)))

MemoryError: Unable to allocate 62.9 GiB for an array with shape (556084, 15179) and data type float64

- degree가 3일 때 degree가 2일때 보다 mse와 rmse가 크다

#### Ridge 클래스 이용

In [13]:
from sklearn.linear_model import Ridge
from sklearn.model_selection import cross_val_score

In [14]:
# 타겟, 피처 데이터 설정
y_target = data['rent_cnt']  # 타겟 데이터
X_data = data.drop(['rent_cnt'],axis=1,inplace=False) # 피처 데이터

In [15]:
ridge = Ridge(alpha=10)
neg_mse_scores = cross_val_score(ridge, X_data, y_target, scoring='neg_mean_squared_error', cv=5) # mse가 음수로 반환
rmse_scores = np.sqrt(-1*neg_mse_scores) # mse를 양수로 변환해 rmse 계산
avg_rmse = np.mean(rmse_scores)

In [16]:
print(' 5 folds 의 개별 Negative MSE scores: ', np.round(neg_mse_scores, 3))
print(' 5 folds 의 개별 RMSE scores : ', np.round(rmse_scores,3))
print(' 5 folds 의 평균 RMSE : {0:.3f} '.format(avg_rmse))

 5 folds 의 개별 Negative MSE scores:  [-0.328 -0.714 -1.163 -1.487 -0.512]
 5 folds 의 개별 RMSE scores :  [0.573 0.845 1.078 1.219 0.715]
 5 folds 의 평균 RMSE : 0.886 


#### alpha 값을 0, 0.1, 1, 10, 100으로 변경하면서 측정

In [17]:
# alpha 파라미터 값 정의
alphas = [0,0.1,1.,10,100]

for alpha in alphas :
    ridge = Ridge(alpha=alpha)
    neg_mse_scores = cross_val_score(ridge, X_data, y_target, scoring='neg_mean_squared_error', cv=5) # mse가 음수로 반환
    rmse_scores = np.sqrt(-1*neg_mse_scores) # mse를 양수로 변환해 rmse 계산
    avg_rmse = np.mean(rmse_scores)
    print('alpha {0} 일 때 5 folds 의 평균 RMSE : {1:.3f} '.format(alpha,avg_rmse))

alpha 0 일 때 5 folds 의 평균 RMSE : 0.886 
alpha 0.1 일 때 5 folds 의 평균 RMSE : 0.886 
alpha 1.0 일 때 5 folds 의 평균 RMSE : 0.886 
alpha 10 일 때 5 folds 의 평균 RMSE : 0.886 
alpha 100 일 때 5 folds 의 평균 RMSE : 0.886 


#### 결과
- 평균 RMSE값이 변화가 없음

### 라쏘 회귀

In [18]:
from sklearn.linear_model import Lasso, ElasticNet

# 뒤에 ElasticNet에서도 사용할 수 있도록 함수로 작성
# alpha값에 따른 회귀 모델의 폴드 평균 RMSE를 출력하고
# 회귀 계수값들을 DataFrame으로 반환하는 함수 생성

def get_linear_reg_eval(model_name,params=None,X_data_n=None,y_target_n=None,verbose=True) :
    coeff_df = pd.DataFrame()
    if verbose : print('####### ', model_name , '#######') # Log 출력
    print(params)    
    for param in params : # alpha값 변화
        # 모델별 해당 클래스 객체 생성
        if model_name =='Ridge': model = Ridge(alpha=param)
        elif model_name =='Lasso': model = Lasso(alpha=param)
        elif model_name =='ElasticNet': model = ElasticNet(alpha=param, l1_ratio=0.7)
            
        # 교차 검증 수행
        neg_mse_scores = cross_val_score(model, X_data_n, 
                                             y_target_n, scoring="neg_mean_squared_error", cv = 5)
        avg_rmse = np.mean(np.sqrt(-1 * neg_mse_scores))
        print('alpha {0}일 때 5 폴드 세트의 평균 RMSE: {1:.3f} '.format(param, avg_rmse))
        
        # cross_val_score는 evaluation metric만 반환하므로 모델을 다시 학습하여 회귀 계수 추출
        model.fit(X_data, y_target_n)
        
        # alpha에 따른 피처별 회귀 계수를 Series로 변환하고 이를 DataFrame의 컬럼으로 추가. 
        coeff = pd.Series(data=model.coef_ , index=X_data.columns )
        colname='alpha:'+str(param)
        coeff_df[colname] = coeff
        
    return coeff_df

In [ ]:
from sklearn.linear_model import Ridge
from sklearn.model_selection import cross_val_score

In [20]:
# 타겟, 피처 데이터 설정
y_target = data['rent_cnt']  # 타겟 데이터
X_data = data.drop(['rent_cnt'],axis=1,inplace=False) # 피처 데이터

In [21]:
lasso = Lasso(alpha=10)
neg_mse_scores = cross_val_score(lasso, X_data, y_target, scoring='neg_mean_squared_error', cv=5) # mse가 음수로 반환
rmse_scores = np.sqrt(-1*neg_mse_scores) # mse를 양수로 변환해 rmse 계산
avg_rmse = np.mean(rmse_scores)

In [22]:
print(' 5 folds 의 개별 Negative MSE scores: ', np.round(neg_mse_scores, 3))
print(' 5 folds 의 개별 RMSE scores : ', np.round(rmse_scores,3))
print(' 5 folds 의 평균 RMSE : {0:.3f} '.format(avg_rmse))

 5 folds 의 개별 Negative MSE scores:  [-0.58  -0.763 -1.438 -1.895 -0.581]
 5 folds 의 개별 RMSE scores :  [0.761 0.873 1.199 1.377 0.762]
 5 folds 의 평균 RMSE : 0.995 


In [28]:
lasso_alphas = [0.02, 0.07, 0.1, 0.5, 1, 3,100]
coeff_lasso_df = get_linear_reg_eval('Lasso',params=lasso_alphas, X_data_n = X_data, y_target_n=y_target)

#######  Lasso #######
[0.02, 0.07, 0.1, 0.5, 1, 3, 100]
alpha 0.02일 때 5 폴드 세트의 평균 RMSE: 0.916 
alpha 0.07일 때 5 폴드 세트의 평균 RMSE: 0.944 
alpha 0.1일 때 5 폴드 세트의 평균 RMSE: 0.952 
alpha 0.5일 때 5 폴드 세트의 평균 RMSE: 0.995 
alpha 1일 때 5 폴드 세트의 평균 RMSE: 0.995 
alpha 3일 때 5 폴드 세트의 평균 RMSE: 0.995 
alpha 100일 때 5 폴드 세트의 평균 RMSE: 0.995 


#### 결과
- alpha값이 0.01, 0.07일 때 가장 좋은 평균 35.357이 나타남

In [29]:
# alpha 값에 따른 피처별 회귀 계수
# 반환된 coeff_lasso_df를 첫번째 컬럼순으로 내림차순 정렬하여 회귀계수 DataFrame출력
coeff_lasso_df

alpha:0.02  alpha:0.07  alpha:0.1  alpha:0.5  alpha:1  alpha:3  \
dust       -0.020579   -0.000000  -0.000000       -0.0     -0.0     -0.0   
temp        0.255641    0.212160   0.182334        0.0      0.0      0.0   
road_cnt    0.000000    0.006388   0.000000        0.0      0.0      0.0   
road_len    0.086963    0.016855   0.000000        0.0      0.0      0.0   
rain       -0.027361   -0.000000  -0.000000       -0.0     -0.0     -0.0   
강남구        -0.214240   -0.000000  -0.000000       -0.0     -0.0     -0.0   
강동구        -0.000000   -0.000000  -0.000000       -0.0     -0.0     -0.0   
강북구        -0.000000   -0.000000  -0.000000       -0.0     -0.0     -0.0   
강서구         0.039318    0.000000   0.000000        0.0      0.0      0.0   
관악구         0.000000    0.000000   0.000000        0.0      0.0      0.0   
광진구         0.000000    0.000000   0.000000        0.0      0.0      0.0   
구로구         0.000000   -0.000000  -0.000000       -0.0     -0.0     -0.0   
금천구        -0.000000   -0.000000  -0.000000       -0.0     -0.0     -0.0   
노원구         0.000000    0.000000   0.000000        0.0      0.0      0.0   
도봉구        -0.000000   -0.000000  -0.000000       -0.0     -0.0     -0.0   
동대문구        0.000000    0.000000   0.000000        0.0      0.0      0.0   
동작구        -0.000000   -0.000000  -0.000000       -0.0     -0.0     -0.0   
마포구         0.000000    0.000000   0.000000        0.0      0.0      0.0   
서대문구       -0.000000   -0.000000  -0.000000       -0.0     -0.0     -0.0   
서초구        -0.000000   -0.000000  -0.000000       -0.0     -0.0     -0.0   
성동구         0.000000    0.000000   0.000000        0.0      0.0      0.0   
성북구         0.000000   -0.000000  -0.000000       -0.0     -0.0     -0.0   
송파구        -0.000000    0.000000   0.000000        0.0      0.0      0.0   
양천구         0.000000    0.000000   0.000000        0.0      0.0      0.0   
영등포구        0.000000    0.000000   0.000000        0.0      0.0      0.0   
용산구        -0.000000   -0.000000  -0.000000       -0.0     -0.0     -0.0   
은평구        -0.000000   -0.000000  -0.000000       -0.0     -0.0     -0.0   
종로구        -0.000000   -0.000000  -0.000000       -0.0     -0.0     -0.0   
중구         -0.000000   -0.000000  -0.000000       -0.0     -0.0     -0.0   
중랑구        -0.000000   -0.000000  -0.000000       -0.0     -0.0     -0.0   
나쁨         -0.000000   -0.000000  -0.000000       -0.0     -0.0     -0.0   
매우나쁨        0.000000    0.000000   0.000000        0.0      0.0      0.0   
보통          0.000000    0.000000   0.000000        0.0      0.0      0.0   
맑음          0.118052    0.000000   0.000000        0.0      0.0      0.0   
비          -0.000000   -0.000000  -0.000000       -0.0     -0.0     -0.0   
폭우         -0.003986   -0.000000  -0.000000       -0.0     -0.0     -0.0   
평일          0.005960    0.000000   0.000000        0.0      0.0      0.0   
주말         -0.000000   -0.000000  -0.000000       -0.0     -0.0     -0.0   
공휴일        -0.000000   -0.000000  -0.000000       -0.0     -0.0     -0.0   
봄          -0.065728   -0.000000  -0.000000       -0.0     -0.0     -0.0   
여름          0.000000    0.000000   0.000000        0.0      0.0      0.0   
가을          0.200387    0.000000   0.000000        0.0      0.0      0.0   
겨울         -0.000000   -0.000000  -0.000000       -0.0     -0.0     -0.0   

          alpha:100  
dust           -0.0  
temp            0.0  
road_cnt        0.0  
road_len        0.0  
rain           -0.0  
강남구            -0.0  
강동구            -0.0  
강북구            -0.0  
강서구             0.0  
관악구             0.0  
광진구             0.0  
구로구            -0.0  
금천구            -0.0  
노원구             0.0  
도봉구            -0.0  
동대문구            0.0  
동작구            -0.0  
마포구             0.0  
서대문구           -0.0  
서초구            -0.0  
성동구             0.0  
성북구            -0.0  
송파구             0.0  
양천구             0.0  
영등포구            0.0  
용산구            -0.0  
은평구            -0.0  
종로구            -0.0  
중구             -0.0  
중랑구  

#### 엘라스틱 넷

In [30]:
elastic_alphas = [ 0.07, 0.1, 0.5, 1, 3, 100]
coeff_elastic_df = get_linear_reg_eval('ElasticNet',params=elastic_alphas, X_data_n = X_data, y_target_n=y_target)
#l1_ratio는 0.7

#######  ElasticNet #######
[0.07, 0.1, 0.5, 1, 3, 100]
alpha 0.07일 때 5 폴드 세트의 평균 RMSE: 0.935 
alpha 0.1일 때 5 폴드 세트의 평균 RMSE: 0.945 
alpha 0.5일 때 5 폴드 세트의 평균 RMSE: 0.995 
alpha 1일 때 5 폴드 세트의 평균 RMSE: 0.995 
alpha 3일 때 5 폴드 세트의 평균 RMSE: 0.995 
alpha 100일 때 5 폴드 세트의 평균 RMSE: 0.995 


#### 결과
- alpha값이 0.07, 0.1일 때 가장 좋은 평균 35.355가 나타남
- 엘라스틱 넷을 이용했을 때 RMSE가 가장 낮음

### 자치구와 기온을 요일과 적용

In [42]:
data2 = data
data2

station_num  date  rent_cnt  gu  dust  rain_seoul  weather  day  temp  \
1               101     0         7  12    30         0.0        0    3  -4.2   
2               101     1        13  12    36         0.0        0    2  -5.0   
3               101     2         9  12    41         0.0        0    2  -5.6   
4               101     3         7  12    44         0.0        0    1  -3.5   
5               101     4        15  12    28         0.0        0    1  -5.5   
...             ...   ...       ...  ..   ...         ...      ...  ...   ...   
794402         4869   344         1  17    32         0.0        0    1  -7.6   
794403         4869   345         4  17    51         0.0        0    1  -4.1   
794404         4869   346         6  17    61         0.2        1    1   0.4   
794405         4869   347         7  17    28         0.0        0    1  -3.9   
794406         4869   348         3  17    22         0.0        0    1  -6.7   

        road_cnt  road_len  rain_gu  
1           57.0      65.7      0.0  
2           57.0      65.7      0.0  
3           57.0      65.7      0.0  
4           57.0      65.7      0.0  
5           57.0      65.7      0.0  
...          ...       ...      ...  
794402     186.0      97.1      0.0  
794403     186.0      97.1      0.0  
794404     186.0      97.1      0.0  
794405     186.0      97.1      0.0  
794406     186.0      97.1      0.0  

[793691 rows x 12 columns]

In [43]:
data2 = data2.drop(['date','station_num','dust','road_cnt', 'road_len', 'rain_seoul', 'weather', 'rain_gu'], axis=1)
data2

rent_cnt  gu  day  temp
1              7  12    3  -4.2
2             13  12    2  -5.0
3              9  12    2  -5.6
4              7  12    1  -3.5
5             15  12    1  -5.5
...          ...  ..  ...   ...
794402         1  17    1  -7.6
794403         4  17    1  -4.1
794404         6  17    1   0.4
794405         7  17    1  -3.9
794406         3  17    1  -6.7

[793691 rows x 4 columns]

### 단순 선형회귀

In [44]:
# 특징과 타겟으로 나누기

x = data2.drop(['rent_cnt'],axis=1)
y = data2[['rent_cnt']]

In [45]:
# 학습과 테스트 분리
X_train, X_test, y_train, y_test = train_test_split(x, y,
                                                   test_size=0.3, random_state=11)

In [46]:
from sklearn.linear_model import LinearRegression

# 빈 선형회귀 모델 생성
line_fitter = LinearRegression() 

# 학습 수행
line_fitter.fit(X_train, y_train)

LinearRegression()

In [47]:
# 학습이 완료된 모델 객체를 활용하여 예측 수행 
pred = line_fitter.predict(X_test)

In [48]:
from sklearn.metrics import mean_squared_error , r2_score

mse = mean_squared_error(y_test, pred)
rmse = np.sqrt(mse) # 사이킷런이 지원x

print('MSE : {0:.3f} , RMSE : {1:.3F}'.format(mse , rmse))
print('Variance score : {0:.3f}'.format(r2_score(y_test, pred)))

MSE : 1160.818 , RMSE : 34.071
Variance score : 0.082


#### 회귀계수와 절편값 확인

In [49]:
print('절편값: ', line_fitter.intercept_)
print('회귀계수값', np.round(line_fitter.coef_, 1))

절편값:  [20.94054054]
회귀계수값 [[-0.1 -2.5  1. ]]


In [50]:
# 5개의 폴드 세트에 교차 검증 수행
# MSE와 RMSE 측정

from sklearn.model_selection import cross_val_score

y_target = data2['rent_cnt']
X_data = data2.drop(['rent_cnt'],axis=1,inplace=False)

lr = LinearRegression()

# 5 fold 셋으로 neg_MSE를 구함
neg_mse_scores = cross_val_score(lr, X_data, y_target, scoring='neg_mean_squared_error', cv=5)

In [51]:
neg_mse_scores

array([ -399.51800973, -1015.15729947, -1624.80184517, -2214.04723203,
        -733.5930169 ])

In [52]:
# RMSE : neg-MSE에 -1  곱하고  sqrt()

rmse_scores = np.sqrt(-1*neg_mse_scores)
avg_rmse = np.mean(rmse_scores)

print(' 5 folds 의 개별 Negative MSE scores: ', np.round(neg_mse_scores, 2))
print(' 5 folds 의 개별 RMSE scores : ', np.round(rmse_scores, 2))
print(' 5 folds 의 평균 RMSE : {0:.3f} '.format(avg_rmse))

 5 folds 의 개별 Negative MSE scores:  [ -399.52 -1015.16 -1624.8  -2214.05  -733.59]
 5 folds 의 개별 RMSE scores :  [19.99 31.86 40.31 47.05 27.08]
 5 folds 의 평균 RMSE : 33.259 


### 다항회귀

In [53]:
y_target = data2['rent_cnt']
X_data = data2.drop(['rent_cnt'],axis=1,inplace=False)

X_train , X_test , y_train , y_test = train_test_split(X_data , y_target ,test_size=0.3, random_state=156)

In [54]:
#(1) degree=2
# - 파이프라인 작업 구성
# Pipeline([('작업명',작업1), ('작업명',작업2),....('작업명',작업n)])
p_model = Pipeline([
    ('poly', PolynomialFeatures(degree=2)),
    ('linear', LinearRegression())
])

# pipeline에 data를 전달 해서 학습
p_model.fit(X_train, y_train)

Pipeline(steps=[('poly', PolynomialFeatures()), ('linear', LinearRegression())])

In [55]:
# linear스텝의 속성 coef_를 반환
print('Polynomial 회귀 계수\n', np.round(p_model.named_steps['linear'].coef_, 2))

Polynomial 회귀 계수
 [ 0.000e+00  1.190e+00 -1.003e+01  8.600e-01 -5.000e-02  3.000e-02
 -0.000e+00  1.980e+00  2.000e-02  1.000e-02]


In [56]:
### 학습된 회귀식을 이용해 예측/평가
y_preds = p_model.predict(X_test)

# 평가
mse = mean_squared_error(y_test, y_preds)
rmse = np.sqrt(mse)

print('MSE : {0:.3f} , RMSE : {1:.3F}'.format(mse , rmse))
print('Variance score : {0:.3f}'.format(r2_score(y_test, y_preds)))

MSE : 1174.420 , RMSE : 34.270
Variance score : 0.086


#### 다항회귀 dgree=2 결과
- MSE, RMSE, variance score가 단항회귀보다 높게 나옴

In [57]:
#(2) degree=2, include_bias 튜닝
# - 파이프라인 작업 구성
# Pipeline([('작업명',작업1), ('작업명',작업2),....('작업명',작업n)])
p_model = Pipeline([
    ('poly', PolynomialFeatures(degree=2, include_bias=False)),
    ('linear', LinearRegression())
])

# pipeline에 data를 전달 해서 학습
p_model.fit(X_train, y_train)

### 학습된 회귀식을 이용해 예측/평가
y_preds = p_model.predict(X_test)

# 평가
mse = mean_squared_error(y_test, y_preds)
rmse = np.sqrt(mse)


print('MSE : {0:.3f} , RMSE : {1:.3F}'.format(mse , rmse))
print('Variance score : {0:.3f}'.format(r2_score(y_test, y_preds)))

MSE : 1174.420 , RMSE : 34.270
Variance score : 0.086


In [58]:
## 3차 다항식으로 변경
# degree=3, include_bias=False
p_model = Pipeline([
    ('poly', PolynomialFeatures(degree=3, include_bias=False)),
    ('linear', LinearRegression())
])

# pipeline에 data를 전달 해서 학습
p_model.fit(X_train, y_train)

### 학습된 회귀식을 이용해 예측/평가
y_preds = p_model.predict(X_test)

# 평가
mse = mean_squared_error(y_test, y_preds)
rmse = np.sqrt(mse)


print('MSE : {0:.3f} , RMSE : {1:.3F}'.format(mse , rmse))
print('Variance score : {0:.3f}'.format(r2_score(y_test, y_preds)))

MSE : 1172.182 , RMSE : 34.237
Variance score : 0.087


- MSE, RMSE는 작아졌지만 Variance score는 커짐

#### Ridge 클래스 이용

In [59]:
from sklearn.linear_model import Ridge
from sklearn.model_selection import cross_val_score

In [60]:
# 타겟, 피처 데이터 설정
y_target = data2['rent_cnt']  # 타겟 데이터
X_data = data2.drop(['rent_cnt'],axis=1,inplace=False) # 피처 데이터

In [61]:
ridge = Ridge(alpha=10)
neg_mse_scores = cross_val_score(ridge, X_data, y_target, scoring='neg_mean_squared_error', cv=5) # mse가 음수로 반환
rmse_scores = np.sqrt(-1*neg_mse_scores) # mse를 양수로 변환해 rmse 계산
avg_rmse = np.mean(rmse_scores)

In [62]:
print(' 5 folds 의 개별 Negative MSE scores: ', np.round(neg_mse_scores, 3))
print(' 5 folds 의 개별 RMSE scores : ', np.round(rmse_scores,3))
print(' 5 folds 의 평균 RMSE : {0:.3f} '.format(avg_rmse))

 5 folds 의 개별 Negative MSE scores:  [ -399.518 -1015.157 -1624.802 -2214.047  -733.593]
 5 folds 의 개별 RMSE scores :  [19.988 31.862 40.309 47.054 27.085]
 5 folds 의 평균 RMSE : 33.259 


#### alpha 값 변경하면서 측정

In [63]:
# alpha 파라미터 값 정의
alphas = [0,0.1,1.,10,100]

for alpha in alphas :
    ridge = Ridge(alpha=alpha)
    neg_mse_scores = cross_val_score(ridge, X_data, y_target, scoring='neg_mean_squared_error', cv=5) # mse가 음수로 반환
    rmse_scores = np.sqrt(-1*neg_mse_scores) # mse를 양수로 변환해 rmse 계산
    avg_rmse = np.mean(rmse_scores)
    print('alpha {0} 일 때 5 folds 의 평균 RMSE : {1:.3f} '.format(alpha,avg_rmse))

alpha 0 일 때 5 folds 의 평균 RMSE : 33.259 
alpha 0.1 일 때 5 folds 의 평균 RMSE : 33.259 
alpha 1.0 일 때 5 folds 의 평균 RMSE : 33.259 
alpha 10 일 때 5 folds 의 평균 RMSE : 33.259 
alpha 100 일 때 5 folds 의 평균 RMSE : 33.259 


- RMSE 값 변화 없음

#### 라쏘 회귀

In [64]:
from sklearn.linear_model import Lasso, ElasticNet

# 뒤에 ElasticNet에서도 사용할 수 있도록 함수로 작성
# alpha값에 따른 회귀 모델의 폴드 평균 RMSE를 출력하고
# 회귀 계수값들을 DataFrame으로 반환하는 함수 생성

def get_linear_reg_eval(model_name,params=None,X_data_n=None,y_target_n=None,verbose=True) :
    coeff_df = pd.DataFrame()
    if verbose : print('####### ', model_name , '#######') # Log 출력
    print(params)    
    for param in params : # alpha값 변화
        # 모델별 해당 클래스 객체 생성
        if model_name =='Ridge': model = Ridge(alpha=param)
        elif model_name =='Lasso': model = Lasso(alpha=param)
        elif model_name =='ElasticNet': model = ElasticNet(alpha=param, l1_ratio=0.7)
            
        # 교차 검증 수행
        neg_mse_scores = cross_val_score(model, X_data_n, 
                                             y_target_n, scoring="neg_mean_squared_error", cv = 5)
        avg_rmse = np.mean(np.sqrt(-1 * neg_mse_scores))
        print('alpha {0}일 때 5 폴드 세트의 평균 RMSE: {1:.3f} '.format(param, avg_rmse))
        
        # cross_val_score는 evaluation metric만 반환하므로 모델을 다시 학습하여 회귀 계수 추출
        model.fit(X_data, y_target_n)
        
        # alpha에 따른 피처별 회귀 계수를 Series로 변환하고 이를 DataFrame의 컬럼으로 추가. 
        coeff = pd.Series(data=model.coef_ , index=X_data.columns )
        colname='alpha:'+str(param)
        coeff_df[colname] = coeff
        
    return coeff_df

In [ ]:
# 타겟, 피처 데이터 설정
y_target = data2['rent_cnt']  # 타겟 데이터
X_data = data2.drop(['rent_cnt'],axis=1,inplace=False) # 피처 데이터

In [ ]:
ridge = Ridge(alpha=10)
neg_mse_scores = cross_val_score(ridge, X_data, y_target, scoring='neg_mean_squared_error', cv=5) # mse가 음수로 반환
rmse_scores = np.sqrt(-1*neg_mse_scores) # mse를 양수로 변환해 rmse 계산
avg_rmse = np.mean(rmse_scores)

In [ ]:
print(' 5 folds 의 개별 Negative MSE scores: ', np.round(neg_mse_scores, 3))
print(' 5 folds 의 개별 RMSE scores : ', np.round(rmse_scores,3))
print(' 5 folds 의 평균 RMSE : {0:.3f} '.format(avg_rmse))

In [65]:
lasso_alphas = [0.02, 0.07, 0.1, 0.5, 1, 3,100]
coeff_lasso_df = get_linear_reg_eval('Lasso',params=lasso_alphas, X_data_n = X_data, y_target_n=y_target)

#######  Lasso #######
[0.02, 0.07, 0.1, 0.5, 1, 3, 100]
alpha 0.02일 때 5 폴드 세트의 평균 RMSE: 33.260 
alpha 0.07일 때 5 폴드 세트의 평균 RMSE: 33.260 
alpha 0.1일 때 5 폴드 세트의 평균 RMSE: 33.261 
alpha 0.5일 때 5 폴드 세트의 평균 RMSE: 33.279 
alpha 1일 때 5 폴드 세트의 평균 RMSE: 33.283 
alpha 3일 때 5 폴드 세트의 평균 RMSE: 33.294 
alpha 100일 때 5 폴드 세트의 평균 RMSE: 35.313 


- 라쏘는 alpha가 0.02, 0.07일 때 가장 낮음
- 하지만 릿지 보단 큼

In [66]:
# alpha 값에 따른 피처별 회귀 계수
# 반환된 coeff_lasso_df를 첫번째 컬럼순으로 내림차순 정렬하여 회귀계수 DataFrame출력
coeff_lasso_df

alpha:0.02  alpha:0.07  alpha:0.1  alpha:0.5   alpha:1   alpha:3  \
gu     -0.055120   -0.054174  -0.053606  -0.046036 -0.036616 -0.000000   
day    -2.384444   -2.221721  -2.124087  -0.822301 -0.000000 -0.000000   
temp    0.965509    0.965106   0.964864   0.961640  0.957317  0.938822   

      alpha:100  
gu    -0.000000  
day   -0.000000  
temp   0.039914

#### 엘라스틱 넷

In [68]:
elastic_alphas = [0.01, 0.07, 0.1, 0.5, 1, 3, 100]
coeff_elastic_df = get_linear_reg_eval('ElasticNet',params=elastic_alphas, X_data_n = X_data, y_target_n=y_target)
#l1_ratio는 0.7

#######  ElasticNet #######
[0.01, 0.07, 0.1, 0.5, 1, 3, 100]
alpha 0.01일 때 5 폴드 세트의 평균 RMSE: 33.259 
alpha 0.07일 때 5 폴드 세트의 평균 RMSE: 33.259 
alpha 0.1일 때 5 폴드 세트의 평균 RMSE: 33.260 
alpha 0.5일 때 5 폴드 세트의 평균 RMSE: 33.271 
alpha 1일 때 5 폴드 세트의 평균 RMSE: 33.283 
alpha 3일 때 5 폴드 세트의 평균 RMSE: 33.291 
alpha 100일 때 5 폴드 세트의 평균 RMSE: 34.984 


- alpha가 0.01, 0.07일 때 평균 RMSE가 33.259로 가장 작음


### 결과
- 자치구를 고려하지 않았을 때가 자치구를 고려했을 때 보다 RMSE값이 높게 나옴
    - 자치구를 고려하지 않았을 때 가장 낮은 RMSE는 alpha가 0.07일 때 35.355임
- 자치구와 기온을 고려했을 때 릿지를 이용하거나 엘라스틱 넷의 alpha 값이 0.07일 때
    - 5 폴드 세트의 RMSE는 33.259로 가장 낮음